### PR TITLE
P2P updater

### PR DESCRIPTION
**Pull request by team-pct**
##P2P updater

Changes proposed in this pull request:
- Update Updater.js
- Add UpdateSeed on Settings
- Add Webtorrent on package.json

I've read the [Contributor License Agreement](/CONTRIBUTING.md#contributor-license-agreement)
* P2P updater

### CODE REVIEW

File: src/app/global.js [hunk 1]
```diff
@@ -57,6 +57,6 @@ var
 
     // Torrent engines
     peerflix = require('peerflix'),
-
+    WebTorrent = require('webtorrent'),
```
Comment (xaiki) [inline src/app/global.js]: we don't need this as a global, please keep it in updater.js

File: src/app/settings.js [hunk 2]
```diff
@@ -149,6 +149,7 @@ Settings.tmpLocation = path.join(os.tmpDir(), Settings.projectName);
 Settings.databaseLocation = path.join(data_path, 'data');
 Settings.deleteTmpOnClose = true;
 Settings.automaticUpdating = true;
+Settings.UpdateSeed = true;
```
Comment (xaiki) [inline src/app/settings.js]: either `Settings.update = {seed: true}`, or Settings.seedUpdate

File: src/app/global.js (hunk 1 above)
Comment (team-pct) [inline src/app/global.js]: We will need it with the new streamer running on Webtorrent Please see #424  #470

**Comment by team-pct** [comment]
I am not sure what you mean by changing the updater behavior but we just included web torrent and kept the Flow of the current updater

File: src/app/global.js (hunk 1 above)
Comment (xaiki) [inline src/app/global.js]: we can change that when it gets merged.

File: src/app/settings.js (hunk 2 above)
Comment (xaiki) [inline src/app/settings.js]: it's just a name nitpick 'seedUpdate' isn't really clear to me.

**Comment by xaiki** [comment]
your PR changed a lot of things (more than can be reviewed) in `updater.js`, let's talk on IRC/Slack and find a way to make this mergeable.

File: src/app/settings.js (hunk 2 above)
Comment (vankasteelj) [inline src/app/settings.js]: naming convention is bad, you should use seedUpdate or redistributeUpdater, but don't start with an uppercase :)

File: package.json [hunk 3]
```diff
@@ -60,6 +60,7 @@
         "nodecast-js": "^1.0.1",
         "opensubtitles-api": "^3.1.0",
         "peerflix": "^0.35.0",
+        "webtorrent": "0.97.2",
```
Comment (vankasteelj) [inline package.json]: better use 0.x.x or ^0.97.2

File: src/app/templates/settings-container.tpl [hunk 4]
```diff
@@ -556,8 +556,12 @@
                 <label class="settings-label" for="cb5"><%= i18n.__("Activate automatic updating") %></label>
             </span>
             <span>
-                <input class="settings-checkbox" name="events" id="cb6" type="checkbox" <%=(Settings.events? "checked='checked'":"")%>>
-                <label class="settings-label" for="cb6"><%= i18n.__("Celebrate various events") %></label>
+                <input class="settings-checkbox" name="UpdateSeed" id="cb6" type="checkbox" <%=(Settings.UpdateSeed? "checked='checked'":"")%>>
+                <label class="settings-label" for="cb6"><%= i18n.__("Activate Update Seeding") %></label>
```
Comment (vankasteelj) [inline src/app/templates/settings-container.tpl]: It's personnal and doesn't really impair the merge of this, but I don't like the string. For starter, you don't need the word "activate" in it, the checkbox says that already. 
How about: 
- "Seed update files to other users"
- "Redistribute the update packages"

File: src/app/updater.js [hunk 5]
```diff
@@ -3,6 +3,8 @@
 
     var CHANNELS = ['stable', 'beta', 'nightly'],
         FILENAME = 'package.nw.new',
+        WebTorrent = require('webtorrent'),
+        client = new WebTorrent({dht: true,maxConns: '5'}),
```
Comment (vankasteelj) [inline src/app/updater.js]: Is the maxConns needed? It fallsback to 55, why forcing a lower number?

File: src/app/updater.js [hunk 6]
```diff
@@ -77,21 +79,44 @@
                 self.updateData = updateData;
                 return true;
             }
+            if (App.settings.UpdateSeed) {
+                client.add(updateData.UpdateUrl, { path: os.tmpdir() }, function (torrent) {
+                    torrent.on('error', function (err) {
+                        win.debug('ERROR' + err.message);
+                    });
+                    torrent.on('done', function () {
+                        win.debug('Seeding the Current Update!');
```
Comment (vankasteelj) [inline src/app/updater.js]: Once 'done', it doesn't seed, according to my tests. You need to explicitly set it with the .seed() api call (I think?)

File: src/app/updater.js [hunk 7]
```diff
@@ -77,21 +79,44 @@
...
+            win.debug('ERROR: ' + err.message);
+            defer.reject(err);
         });
+
+        client.add(source, {
+            path: outputDir
+        }, function (torrent) {
+            win.debug('Downloading update... Please allow a few minutes');
+            torrent.on('error', function (err) {
+                win.debug('ERROR' + err.message);
```
Comment (vankasteelj) [inline src/app/updater.js]: Shouldn't it be `win.warn('Downloading update failed:', error)` ? Make it visible but not a 'breaking' error, and also comprehensible.

File: src/app/updater.js [hunk 8]
```diff
@@ -77,21 +79,44 @@
...
+                defer.reject(err);
+            });
+            torrent.on('done', function () {
+                win.debug('Update downloaded!');
+                defer.resolve(path.join(outputDir, torrent.name));
```
Comment (vankasteelj) [inline src/app/updater.js]: torrent.name contains the extension of the file in this case? Because there's a risk it's a directory

Comment (xaiki) [inline src/app/updater.js]: we create the update torrent.

File: src/app/updater.js (hunk 6 above)
Comment (xaiki) [inline src/app/updater.js]: so it looks like seed() is to create a new torrent, i'm not sure how seed control works...

**Comment by xaiki** [comment]
closing as #545 rebases it

**Pull request closed without merging**